src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1888,8 +1894,18 @@ fn specialize<'p, 'a: 'p, 'tcx>(
             }
         }
 
-        PatternKind::Or { .. } => {
-            bug!("support for or-patterns has not been fully implemented yet.");
+        PatternKind::Or { ref pats } => {
+            let mut specialized_pats = vec![];
+            for pat in pats {
+                if let Some(s) = specialize(cx, &[pat], constructor, wild_patterns) {
+                    specialized_pats.extend(s);
+                }
+            }
+            if specialized_pats.is_empty() {
+                None
+            } else {
+                Some(SmallVec::from_vec(specialized_pats))
+            }
```

Comment:
I believe this behaviour is incorrect; you are breaking the invariant that all the rows of the matrix have the same number of rows. You'd need specialize to return a Vec instead of an Option

Reply:
:+1: yeah i hit an error in the update here, so this is not in the update.

Reply:
Hit some snags in the work for this... trying to figure out how to get the otherwise blocks and the current block right so that each block gets the correct single terminator.